packages/xforms-engine/test/parse/xpath/dependency-analysis.test.ts

```diff
@@ -0,0 +1,337 @@
+				'No context; absolute paths in expression; extracted and returned as-is, already resolved',
+
+			contextNodeset: null,
+			expression: '/foo/bar | /bat',
+			expected: ['/foo/bar', '/bat'],
+		},
+
+		{
+			description:
+				'Relative context; absolute paths in expression; extracted and returned as-is, already resolved',
+
+			contextNodeset: 'quux',
+			expression: '/foo/bar and /bat',
+			expected: ['/foo/bar', '/bat'],
+		},
+
+		{
+			description:
+				'Absolute context; absolute paths in expression; extracted and returned as-is, already resolved',
+
+			contextNodeset: '/quux',
+			expression: '/foo/bar or /bat',
+			expected: ['/foo/bar', '/bat'],
+		},
+
+		{
+			description:
+				'No context; relative paths in expression; extracted and returned as-is, no resolution context available',
+
+			contextNodeset: null,
+			expression: 'foo/bar | bat',
+			expected: ['foo/bar', 'bat'],
+		},
+
+		{
+			description:
+				'Relative context; relative paths in expression; extracted and contextualized as relative to the input context',
+
+			contextNodeset: 'quux',
+			expression: 'foo/bar or bat',
+			expected: ['quux/foo/bar', 'quux/bat'],
+		},
+
+		{
+			description:
+				'Absolute context; relative paths in expression; extracted and contextualized as relative to the input context',
+
+			contextNodeset: '/quux',
+			expression: 'foo/bar or bat',
+			expected: ['/quux/foo/bar', '/quux/bat'],
+		},
+
+		{
+			description:
+				'Absolute context; relative paths with abbreviated self and parent steps; extracted, and abbreviated steps resolved, relative to the input context',
+
+			contextNodeset: '/data/grp',
+			expression: './foo/bar and ../quux',
+			expected: ['/data/grp/foo/bar', '/data/quux'],
+		},
+
+		{
+			description:
+				'Absolute context; current() FilterExpr paths (one with a subsequent abbreviated parent step); extracted, current() treated as analogous to self step, parent step resolved, finally relative to the input context',
+
+			contextNodeset: '/data/grp',
+			expression: 'current()/foo/bar and current()/../quux',
+			expected: ['/data/grp/foo/bar', '/data/quux'],
+		},
+
+		{
+			description: 'No context; instance() and relative path; extracted and returned and as-is',
+
+			contextNodeset: null,
+			expression: 'instance("foo")/bar | ./quux',
+			expected: ['instance("foo")/bar', './quux'],
+		},
+
+		{
+			description:
+				'Relative context; instance() and relative path; extracted, instance call returned as-is, relative paths resolved to context',
+
+			contextNodeset: 'sel1',
+			expression: 'instance("foo")/bar and (bat or ./quux)',
+			expected: ['instance("foo")/bar', 'sel1/bat', 'sel1/quux'],
+		},
+
+		{
+			description:
+				'Relative context; instance() and relative path; extracted, instance call returned as-is, relative paths resolved to context',
+
+			contextNodeset: 'sel1',
+			expression: 'instance("foo")/bar and (bat or ../quux)',
+			expected: ['instance("foo")/bar', 'sel1/bat', 'sel1/../quux'],
```

Comment:
'sel1/../quux' -- subtle (in that it can't be `quux`) Can this actually be exercised in a form? Maybe in a repeat? Not super important but it piqued my curiosity.

Reply:
I don't know if I can think of a form off the top of my head.

Connecting this to the subtlety I noticed in the [previous thread](https://github.com/getodk/web-forms/pull/166#discussion_r1684725331), I think both cases could be resolved more completely by moving the hierarchic check into a Predicate:

- I think this case would be semantically equivalent to (say) `quux[count(../sel1) > 0]`
- The other would be equivalent to `//foo[count(.) > 0]`

Moving those to predicates would definitely be more involved, and I'm not inclined to even think about it in this scope. But assuming this broad approach to dependency tracking survives long term (which I'm iffy on), I wouldn't be surprised if edge cases come along where we'd think about an approach like that.

Reply:
> I think this case would be semantically equivalent to (say) quux[count(../sel1) > 0]

Agreed